tools/pipeline-generator/docs/design.md

```diff
@@ -0,0 +1,239 @@
+
+The goal of this project is to build a robust tool that enables a seamless path for developers to integrate AI models exported in the **MONAI Bundle format** into inference-based applications built with the **MONAI Deploy App SDK** and the **Holoscan SDK**.
+
+This tool will support:
+
+- Standard MONAI Bundles (.pt, .ts, .onnx)
+- MONAI Bundles exported in **Hugging Face-compatible format**
+
+By bridging the gap between model packaging and application deployment, this project aims to simplify clinical AI prototyping and deployment across NVIDIA’s edge AI platforms.
+
+## **Background**
+
+The **MONAI Bundle** is a standardized format designed to package deep learning models for medical imaging. It includes the model weights, metadata, transforms, and documentation needed to make the model self-contained and portable.
+
+The **Holoscan SDK** is NVIDIA’s real-time streaming application SDK for AI workloads in healthcare and edge devices. The **MONAI Deploy App SDK** is designed for building composable inference applications, often used in radiology and imaging pipelines.
+
+As of MONAI Core, bundles can also be exported in a **Hugging Face-compatible format**, which allows sharing through the Hugging Face Model Hub. Supporting this format increases reach and adoption.
+
+## **Benefits**
+
+- Speeds up deployment of MONAI-trained models in Holoscan/Deploy pipelines
+- Ensures standardized and reproducible model integration
+- Makes AI development more accessible to healthcare and edge-AI developers
+- Enables the usage of models from Hugging Face directly in clinical-style workflows
+
+## **Assumptions/Limitations**
+
+- The tool does not convert input formats given that each model may expect a different type of input
+- The tool does not convert output formats given that each model may output a different type of result
+- The tool supports only torchscript (ts) models
+
+## **Scope**
+
+This project includes:
+
+- Support for loading and parsing standard MONAI Bundles (P0)
+- Support for Hugging Face-exported MONAI Bundles (P0)
+- Integration with MONAI Deploy App SDK (P0)
+- Dynamic generation of pre/post-processing pipelines from metadata (P0)
+- Integration with Holoscan SDK’s inference operators (P1)
+- Tools to validate and prepare MONAI Bundles for deployment (P1)
+
+## **Key Features**
+
+- **Bundle Parsing Utility**
+  - Parses metadata.json, inference.json, and other relevant files
+  - Extracts model paths, input/output shapes, transform descriptions, and model metadata
+  - Detects format: .pt, .ts, .onnx, or Hugging Face variant
+- **Model Format Support**
+  - TorchScript (.ts): Loaded with torch.jit.load()
+  - ONNX (.onnx): Loaded with ONNXRuntime or TensorRT
+  - PyTorch state dict (.pt): Loaded with model definition code/config
+  - Hugging Face-compatible: Recognized and unpacked with reference to Hugging Face conventions
+- **AI Inference Operator Integration**
+  - Python and C++ support for TorchScript/ONNX-based inference
```

Comment:
So far no C++ in the app code,

Reply:
I believe the idea from @rahul-imaging is to be able to genrate a C++ application, consuming MONAI Bundle, using this tool.